txstratum/manager.py

```diff
@@ -0,0 +1,206 @@
+        self.update_miners_block_template()
+
+    def update_miners_block_template(self) -> None:
+        """Create and send a new job for each miner."""
+        for miner, protocol in self.miners.items():
+            if isinstance(protocol.current_job, MinerBlockJob):
+                self.update_miner_job(protocol)
+
+    def status(self) -> Dict[Any, Any]:
+        """Return status dict with useful metrics for use in Status API."""
+        miners = [p.status() for p in self.miners.values()]
+        total_hashrate_ghs = sum(p.hashrate_ghs or 0 for p in self.miners.values())
+        return {
+            'miners': miners,
+            'total_hashrate_ghs': total_hashrate_ghs,
+            'started_at': self.started_at,
+            'uptime': self.uptime,
+            'tx_queue': len(self.tx_queue),
+            'tx_jobs': [job.to_dict() for job in self.tx_jobs.values()],
+            'block_template': self.block_template.to_dict() if self.block_template else None,
+        }
+
+    def get_job(self, uuid: bytes) -> Optional[MinerTxJob]:
+        """Return the MinerTxJob related to the uuid."""
+        return self.tx_jobs.get(uuid)
+
+    def add_tx(self, tx_bytes: bytes, *, add_parents: bool = False) -> MinerTxJob:
+        """Add new tx to be mined."""
+        # TODO Add parents.
+        uuid = MinerTxJob.get_uuid(tx_bytes)
+        if uuid in self.tx_jobs:
+            return self.tx_jobs[uuid]
+        job = MinerTxJob(tx_bytes)
+        self.tx_jobs[job.uuid] = job
+        self.enqueue_tx_job(job)
+        return job
+
+    def cancel_job(self, job: MinerTxJob) -> None:
+        """Cancel tx mining job."""
+        self.tx_jobs.pop(job.uuid)
+        self.tx_queue.remove(job)
+        self.stop_mining_tx(job)
+
+    def enqueue_tx_job(self, job: MinerTxJob) -> None:
+        """Enqueue a tx job to be mined."""
+        if len(self.tx_queue) > 0:
+            # If the queue is not empty, just put at the end.
+            self.tx_queue.append(job)
+            return
+
+        candidates = [x for x in self.miners.values() if x.is_mining_block()]
+        if len(candidates) == 0:
+            # If all miners busy with transactions, queue it.
+            self.tx_queue.append(job)
+            return
+
+        best_match = max(candidates, key=lambda x: x.current_weight)
+        best_match.update_job(job, clean=True)
+
+    def stop_mining_tx(self, job: MinerTxJob) -> None:
+        """Stop mining a tx job."""
+        for protocol in self.miners.values():
+            if protocol.current_job == job:
+                self.update_miner_job(protocol)
+
+    def _job_clean_up(self, job: MinerTxJob) -> None:
+        """Clean up tx job. It is scheduled after a tx is solved."""
+        self.tx_jobs.pop(job.uuid)
+
+    def get_best_job(self, protocol: StratumProtocol) -> MinerJob:
+        """Return best job for a miner."""
+        if len(self.tx_queue) > 0:
+            return self.tx_queue.popleft()
```

Comment:
I'm not sure this is the exact point where we would implement these changes, but I thought it was best to comment here so we could keep all in a thread (as opposed to a single comment on the PR).

There are 2 things I've seen are important for mining transactions and I think haven't been implemented:
- adding a timeout. In our latest implementation we use 1s;
- update tx timestamp/parents when sending the job to the miner. This ensures miners get new data to mine on;

What we'd see is that sometimes the miners would not find the correct nonce for a transaction, even with low weight (let's say 20) in a reasonable amount of time. That's why it's important to have the timeout and to update the tx, otherwise the miner would mine over the same piece of data.

Reply:
I've implemented the update timestamp for both tx and blocks. What do you mean by a timeout? We already update the job every 30 seconds, updating both the timestamp and the weight. The target time for blocks is 15 seconds.

Should I implement different job update interval? Maybe 30 seconds for blocks and 5 seconds for transactions?

Reply:
We've been using 1 second timeouts for tx jobs. We had a discussion here: https://gitlab.com/HathorNetwork/hathor-python/-/merge_requests/398#note_339321295

Reply:
I implemented a 2 second update. What do you think?

Reply:
2 seconds seems good for now. We might want to have an adjustable (per miner) interval in the future. When I did some testing in our stratum, my miner was better with 1s intervals (the internet here is pretty reliable). But I understand Jan had a bigger latency, so 1s might be too quick for him.